Dual stack healthz server

aroradaman opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug
/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #116486

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
kube-proxy extends the schema of its healthz/ and livez/ endpoints to incorporate information about the corresponding IP family
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

aroradaman committed:
pkg/proxy/healthcheck: rename 'proxier' to 'proxy'

KubeProxy operates with a single health server and two proxies,
one for each IP family. The use of the term 'proxier' in the
types and functions within pkg/proxy/healthcheck can be
misleading, as it may suggest the existence of two health
servers, one for each IP family.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/healthcheck: consolidate IsHealthy and isHealthy

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/test pull-kubernetes-typecheck
/test pull-kubernetes-e2e-kind 
/test pull-kubernetes-e2e-gce

aroradaman committed:
pkg/proxy/healthcheck: file rename

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman commented:
/assign @aojea @danwinship

danwinship commented on pkg/proxy/healthcheck/proxier_health.go:
```diff
@@ -101,14 +101,7 @@ func (hs *ProxyHealthServer) QueuedUpdate(ipFamily v1.IPFamily) {
-}
-
-func (hs *ProxyHealthServer) isHealthy() (bool, time.Time) {
+func (hs *ProxyHealthServer) Health() (bool, time.Time) {
```
can you keep the comment here (well, update the comment)

danwinship commented on pkg/proxy/healthcheck/healthcheck_test.go:
```diff
@@ -142,8 +142,11 @@ type fakeProxyHealthChecker struct {
+	return ProxyHealthCheckStatus{
+		Healthy:     fake.healthy,
+		LastUpdated: time.Now(),
+	}
```
If the additional fields aren't relevant to the unit tests then explain why. (Does this mean the unit tests aren't testing some functionality?)

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -56,6 +57,29 @@ type ProxyHealthServer struct {
+
+// ProxyHealthCheckStatus represents the health check status of
+// kube-proxy, embeds health check status of individual proxiers.
+type ProxyHealthCheckStatus struct {
```
Maybe just `ProxyHealth`, `ProxierHealth`? (Or `ProxyHealthStatus`, `ProxierHealthStatus`.) To me, "HealthCheckStatus" implies "the status of performing a health check" (eg, "still waiting for a response", "got a response", "timed out")

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -124,9 +155,14 @@ func (hs *ProxyHealthServer) Health() (bool, time.Time) {
+		status.Status[ipFamily] = ProxierHealthCheckStatus{
+			LastUpdated: proxierLastUpdated,
+			Healthy:     false,
+		}
```
It seems ugly to have two separate `status.Status[ipFamily] = ProxierHealthCheckStatus{...}` assignments. Can you rewrite it to just set local variables and then only assign to the map at the end?

aroradaman replied:
This was intentional as we have continue/skip conditions in between and assigning at the end with local variable might not happen every time, will refactor the existing logic to accomodate this.

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -124,9 +155,14 @@ func (hs *ProxyHealthServer) Health() (bool, time.Time) {
 	}
-	return true, lastUpdated
+	status.LastUpdated = lastUpdated
+	return status
```
Alternatively, would it simplify things to just have a `ProxierHealthCheckStatus` field in `ProxyHealthServer`, and update it when `Updated` is called, and then just return it immediately from `Health()`?

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -164,13 +200,13 @@ func (hs *ProxyHealthServer) Run(ctx context.Context) error {
 	listener, err := hs.listener.Listen(ctx, hs.addr)
 	if err != nil {
-		return fmt.Errorf("failed to start proxy healthz on %s: %v", hs.addr, err)
+		return fmt.Errorf("failed to start proxy healthz on %s: %w", hs.addr, err)
```
I suspect the linter is forcing you to do this because you renamed the file, in which case you probably ought to do it as part of the rename commit. (Likewise the `_ *http.Request` below.)

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -200,19 +238,29 @@ func (h healthzHandler) ServeHTTP(resp http.ResponseWriter, req *http.Request) {
 		lastUpdated = currentTime
 	}
-	fmt.Fprintf(resp, `{"lastUpdated": %q,"currentTime": %q, "nodeEligible": %v}`, lastUpdated, currentTime, nodeEligible)
+	output, err := json.Marshal(status)
+	if err != nil {
+		klog.ErrorS(err, "unexpected error serializing JSON")
+		_, _ = fmt.Fprintf(resp, `{"lastUpdated": %q,"currentTime": %q, "nodeEligible": %v}`,
+			lastUpdated, status.CurrentTime, status.NodeEligible)
```
This should not happen so don't bother doing a fallback

aroradaman committed:
pkg/proxy/healthcheck: enhance testing

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
pkg/proxy/healthcheck: bug fix for last updated time

The lastUpdated time returned by healthz call should be the latest
lastUpdated time among the proxiers. Prior to this commit, if proxy
is unhealthy, the returned lastUpdated time was lastUpdated time
of the unhealthy proxier.

Signed-off-by: Daman Arora <aroradaman@gmail.com>

aroradaman committed:
kube-proxy healthz handler ip family aware

Signed-off-by: Daman Arora <aroradaman@gmail.com>
Co-authored-by: Antonio Ojea <aojea@google.com>

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -35,6 +37,29 @@ const (
 	ToBeDeletedTaint = "ToBeDeletedByClusterAutoscaler"
 )
 
+// ProxierHealth represents the health check status of
+// proxier which operates on a single IP family.
+type ProxierHealth struct {
```
```
// ProxierHealth represents the health of a proxier which operates on a single IP family
```

("health", not "health check status", and also, don't wrap the comment so narrowly... k8s code tends to go for about 100 character lines.)

(likewise for ProxyHealth)

(you could rewrap the comments to be wider throughout the new code)

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -125,9 +162,16 @@ func (hs *ProxyHealthServer) Health() (bool, time.Time) {
+
+		// mark the status unhealthy.
+		health.Healthy = false
+		health.Status[ipFamily] = ProxierHealth{
+			LastUpdated: proxierLastUpdated,
+			Healthy:     false,
+		}
```
`health.Status[ipFamily].Healthy = false`
?

aroradaman replied:
golang doesn't allow this, right?
https://github.com/golang/go/issues/3117

```
cannot assign to struct field health.Status[ipFamily].Healthy in map
```

danwinship commented on pkg/proxy/healthcheck/proxy_health.go:
```diff
@@ -181,11 +225,13 @@ type healthzHandler struct {
-	currentTime := h.hs.clock.Now()
+	healthy := health.Healthy && nodeEligible
+	// updating the node eligibility here (outside of Health() call)
+	// as we only want responses of /healthz calls to have that.
```
"as we only want responses of /healthz calls (not /livez) to have that"

danwinship reviewed (commented):
Nice to see that adding more unit tests turned up a bug.

There are linter and gofmt errors to address

danwinship commented:
/lgtm
/approve

aroradaman commented:
/kind feature

aojea commented:
Thanks @aroradaman
